src/collapse/dtrace.rs

```diff
@@ -0,0 +1,219 @@
+pub struct Folder {
+    /// All lines until the next empty line are stack lines.
+    in_event: bool,
+
+    /// Skip all stack lines in this event.
+    skip_stack: bool,
+
+    /// Function entries on the stack in this entry thus far.
+    stack: VecDeque<String>,
+    count: usize,
+    /// Number of times each call stack has been seen.
+    occurrences: HashMap<String, usize>,
+
+    /// Current comm name.
+    ///
+    /// Called pname after original stackcollapse-perf source.
+    pname: Option<String>,
+
+    //    event_filtering: EventFilterState,
+    opt: Options,
+}
+
+impl Collapse for Folder {
+    fn collapse<R, W>(&mut self, mut reader: R, writer: W) -> io::Result<()>
+    where
+        R: BufRead,
+        W: Write,
+    {
+        let mut line = String::new();
+        //TODO: clean this up, we skip the first 4 lines:
+        //        CPU     ID                    FUNCTION:NAME
+        reader.read_line(&mut line)?;
+        //        0  64091                        :tick-60s
+        reader.read_line(&mut line)?;
+        //
+        reader.read_line(&mut line)?;
+        //
+        reader.read_line(&mut line)?;
+        loop {
+            line.clear();
+
+            if reader.read_line(&mut line)? == 0 {
+                break;
+            }
+
+            if line.starts_with('#') {
+                continue;
+            }
+
+            let line = line.trim_end();
+            if line.is_empty() {
+                self.after_event();
+            } else {
+                self.on_line(line);
+            }
+        }
+        self.after_event();
+        self.finish(writer)
+    }
+}
+
+impl From<Options> for Folder {
+    fn from(mut opt: Options) -> Self {
+        Self {
+            in_event: false,
+            skip_stack: false,
+            stack: VecDeque::default(),
+            count: 0,
+            occurrences: HashMap::default(),
+            pname: None,
+            opt,
+        }
+    }
+}
+
+impl Folder {
+    fn on_line(&mut self, line: &str) {
+        self.in_event = true;
+        self.on_stack_line(line)
+    }
+
+    fn stack_line_parts(line: &str) -> Option<(&str, &str)> {
+        let mut line = line.splitn(2, '+');
+        let path = line.next()?;
+
+        let addr = if let Some(addr) = line.next() {
+            addr
+        } else {
+            ""
+        };
+        Some((path, addr))
+    }
+
+    // we have a stack line that shows one stack entry from the preceeding event, like:
+    //
+    //     unix`tsc_gethrtimeunscaled+0x21
+    //     genunix`gethrtime_unscaled+0xa
+    //     genunix`syscall_mstate+0x5d
+    //     unix`sys_syscall+0x10e
+    //       1
+    fn on_stack_line(&mut self, line: &str) {
+        let line = line.trim_start();
+        if let Ok(count) = line.parse::<usize>() {
+            self.count = count;
+        } else {
+            if let Some((func, addr)) = Folder::stack_line_parts(line) {
+                self.stack.push_front(func.to_string());
```

Comment:
This doesn't seem quite right.
First, upstream [only](https://github.com/brendangregg/FlameGraph/blob/9d7c1b1233839de1f842135be84e54f2470adb42/stackcollapse.pl#L85) removes the address part if `includeoffset` isn't set.
Second, it seems to do some [additional post-processing](https://github.com/brendangregg/FlameGraph/blob/9d7c1b1233839de1f842135be84e54f2470adb42/stackcollapse.pl#L87L100) on the function name that we probably also want to do. You should feel free to ignore the [Java un-inlining](https://github.com/brendangregg/FlameGraph/blob/9d7c1b1233839de1f842135be84e54f2470adb42/stackcollapse.pl#L93) for the time being if you want to, though it's [not too bad](https://github.com/jonhoo/inferno/blob/b7605ef69cc53e07612e098264b3bd6499ab5877/src/collapse/perf.rs#L275L281).

Reply:
👍